src/main/java/hudson/plugins/active_directory/ActiveDirectoryAuthenticationProvider.java

```diff
@@ -99,64 +124,84 @@ static String dnToLdapUrl(String dn) {
-            }
-            groups.add(SecurityRealm.AUTHENTICATED_AUTHORITY);
+                        List<GrantedAuthority> groups = new ArrayList<GrantedAuthority>();
+                        for( Com4jObject g : usr.groups() ) {
+                            if (g==null) {
+                                continue;   // according to JENKINS-17357 in some environment the collection contains null
+                            }
+                            IADsGroup grp = g.queryInterface(IADsGroup.class);
+                            // cut "CN=" and make that the role name
+                            groups.add(new GrantedAuthorityImpl(grp.name().substring(3)));
+                        }
+                        groups.add(SecurityRealm.AUTHENTICATED_AUTHORITY);
 
-            LOGGER.log(Level.FINER, "Login successful: "+username+" dn="+dn);
+                        LOGGER.log(Level.FINER, "Login successful: " + username + " dn="+ dn);
 
-            return new ActiveDirectoryUserDetail(
-                username, password,
-                !isAccountDisabled(usr),
-                true, true, true,
-                groups.toArray(new GrantedAuthority[groups.size()]),
-                    getFullName(usr), getEmailAddress(usr), getTelephoneNumber(usr)
-            ).updateUserInfo();
-        } catch (AuthenticationException e) {
-            LOGGER.log(Level.FINE, "Failed toretrieve user "+username, e);
-            throw e;
-        } finally {
-            col.disposeAll();
-            COM4J.removeListener(col);
+                        return new ActiveDirectoryUserDetail(
+                                username, password,
+                                !isAccountDisabled(usr),
+                                true, true, true,
+                                groups.toArray(new GrantedAuthority[groups.size()]),
+                                getFullName(usr), getEmailAddress(usr), getTelephoneNumber(usr)
+                        ).updateUserInfo();
+                    } catch (AuthenticationException e) {
+                        LOGGER.log(Level.FINE, "Failed to retrieve user "+ username, e);
+                        //We need throw the AuthenticationException to re-throw later in UncheckedExecutionException
+                        throw e;
+                    } finally {
+                        col.disposeAll();
+                        COM4J.removeListener(col);
+                    }
+                }
+            });
+        } catch (UncheckedExecutionException e) {
+            Throwable t = e.getCause();
+            if (t instanceof AuthenticationException) {
+                AuthenticationException authenticationException = (AuthenticationException)t;
+                throw authenticationException;
+            }
+        } catch (java.util.concurrent.ExecutionException e) {
+            LOGGER.log(Level.SEVERE, "There was a problem caching user "+ username, e);
         }
+        return userDetails;
```

Comment:
according to the [latest spring docs](http://docs.spring.io/autorepo/docs/spring-security/3.2.6.RELEASE/apidocs/index.html?org/springframework/security/authentication/dao/AbstractUserDetailsAuthenticationProvider.html) (which may have changed since acegi) this must never be `null` - yet you have code paths where this could be the case....